plasmapy/plasma/grids.py

```diff
@@ -1098,96 +1192,130 @@ def volume_averaged_interpolator(
-            persistent = False
-
         # If not persistent, clear the cached properties so they are re-created
         # when called below
         if not persistent:
             try:
                 del self._interp_quantities
+            except AttributeError:
+                pass
+            try:
                 del self._interp_units
             except AttributeError:
                 pass
 
-        # Create a list of empty arrays to hold results
-        sum_value = np.zeros([nparticles, nargs])
-
-        # Strip units from pos (eliminate unit operations in loop)
-        pos = pos.si.value
-
-        # Calculate the grid positions for each particle as interpolated
-        # by the nearest neighbor interpolator
-        xpos = ax0[i[:, 0]]
-        ypos = ax1[i[:, 1]]
-        zpos = ax2[i[:, 2]]
-
-        # Determine the points bounding the grid cell containing the
-        # particle
-        x0 = np.where(pos[:, 0] > xpos, i[:, 0], i[:, 0] - 1)
-        x1 = x0 + 1
-        y0 = np.where(pos[:, 1] > ypos, i[:, 1], i[:, 1] - 1)
-        y1 = y0 + 1
-        z0 = np.where(pos[:, 2] > zpos, i[:, 2], i[:, 2] - 1)
-        z1 = z0 + 1
-
-        # Calculate the cell volume
-        cell_vol = self.dax0.si.value * self.dax1.si.value * self.dax2.si.value
-        n0, n1, n2 = self.shape
+        # -- begin averaging --
 
-        # Go through all of the vertices around the position and volume-
-        # weight the values
-        for x in [x0, x1]:
-            for y in [y0, y1]:
-                for z in [z0, z1]:
+        nparticles = pos.shape[0]
+        nargs = len(args)
 
-                    # Determine if gridpoint is within bounds
-                    valid = (
-                        (x >= 0) & (x < n0) & (y >= 0) & (y < n1) & (z >= 0) & (z < n2)
-                    )
-                    out = np.where(~valid)
+        # Load grid attributes (so this isn't repeated)
+        ax0, ax1, ax2 = self._ax0_si, self._ax1_si, self._ax2_si
+        dx, dy, dz = self._dax0_si, self._dax1_si, self._dax2_si
+        n0, n1, n2 = self.shape
+
+        # find cell nearest to each position
+        nearest_neighbor_index = np.zeros((nparticles, 3), dtype=np.int32)
+        nearest_neighbor_index[..., 0] = np.abs(pos[:, 0, None] - ax0).argmin(axis=1)
+        nearest_neighbor_index[..., 1] = np.abs(pos[:, 1, None] - ax1).argmin(axis=1)
+        nearest_neighbor_index[..., 2] = np.abs(pos[:, 2, None] - ax2).argmin(axis=1)
+
+        # Create a mask for positions that are off the grid. The values at
+        # these points will be set to zero later.
+        mask_particle_off = (
+            (pos[:, 0] < ax0.min() + 0.5 * dx)
+            | (pos[:, 0] > ax0.max() - 0.5 * dx)
+            | (pos[:, 1] < ax1.min() + 0.5 * dy)
+            | (pos[:, 1] > ax1.max() - 0.5 * dy)
+            | (pos[:, 2] < ax2.min() + 0.5 * dz)
+            | (pos[:, 2] > ax2.max() - 0.5 * dz)
+        )
```

Comment:
```suggestion
        mask_particle_off = (
            (pos[:, 0] < ax0.min())
            | (pos[:, 0] > ax0.max())
            | (pos[:, 1] < ax1.min())
            | (pos[:, 1] > ax1.max())
            | (pos[:, 2] < ax2.min())
            | (pos[:, 2] > ax2.max())
        )
```

Reply:
I think the current version of this is correct: as soon as a particle crosses `ax0.max() - dx/2` part of the particle volume is off the grid and therefore undefined.